hubble: Refactor drop event emitter as hive cell

devodev opened the pull request:
Behaviour changes:
- Call Shutdown() on the broadcaster to ensure we stop sending events to the event sink when the hive stops.
  - The broadcaster allows receiving a context and will shutdown itself in a goroutine when it is cancelled, but we don't do that because we need to provide the context upfront, which we can't do from a cell constructor.
- Ensure that ProcessFlow exits if the passed context is cancelled.
 
Fixes: #40061

devodev committed:
hubble: Refactor drop event emitter as hive cell

Call Shutdown() on the broadcaster to ensure we stop sending
events to the event sink when the hive stops.

Ensure that ProcessFlow exits if the passed context is cancelled.

Convert endpointToString and l4protocolToString to functions
as they don't need access to the emitter fields.

Signed-off-by: Alexandre Barone <abalexandrebarone@gmail.com>

kaworu commented on pkg/hubble/dropeventemitter/dropeventemitter.go:
```diff
@@ -39,30 +41,35 @@ func NewDropEventEmitter(interval time.Duration, reasons []string, k8s client.Cl
+	default:
+	}
+
 	reason := strings.ToLower(flow.DropReasonDesc.String())
```
not directly related to the patch but: we should consider `strings.ToLower` the `--hubble-drop-events-reasons` list in `(*config).normalize()` or `new()`. Kind of strange for the Hubble CLI to print "POLICY_DENIED" but don't "recognise" it here (later `slices.Contains(e.reasons, reason)` would return `false` as the flow's reason was `toLower`ed, but not the reasons parsed from the `--hubble-drop-events-reasons` flag).

devodev replied:
Ha interesting. I can certainly add a commit to update the behavior.

devodev replied:
I've updated to perform normalization in the new method. I felt it better to keep required invariants of the type in its constructor instead of outside.
I've also added logic to filter out invalid drop reasons,  because why not!

kaworu reviewed (approved):
Thanks for the cleanup @devodev! :tada: 

Patch LGTM, one non-blocking comment on the way.

devodev committed:
hubble: Normalize drop reasons in dropeventemitter

When creating a dropeventemitter, ensure we validate reasons on
input and store them as enum value instead of strings.

Additionally, ensure that invalid drop reasons are filtered out and
emit a warning log.

Finally, use the drop reason enum when declaring the default config.

Signed-off-by: Alexandre Barone <abalexandrebarone@gmail.com>

kaworu commented on pkg/hubble/dropeventemitter/dropeventemitter.go:
```diff
@@ -13,23 +14,27 @@ import (
 	typedv1 "k8s.io/client-go/kubernetes/typed/core/v1"
 	"k8s.io/client-go/tools/record"
 
+	"github.com/cilium/cilium/api/v1/flow"
```
we already import it as flowpb

devodev replied:
oups

kaworu commented on pkg/hubble/dropeventemitter/dropeventemitter.go:
```diff
@@ -39,30 +44,44 @@ func NewDropEventEmitter(interval time.Duration, reasons []string, k8s client.Cl
+			continue
+		}
+		reasonsNormalized = append(reasonsNormalized, strings.ToLower(r))
+	}
```
If we're getting the "enum" value anyway, let's use it as internal representation instead of strings.
```suggestion
        rs := make([]flowpb.DropReason, 0, len(reasons))
        for _, reason := range reasons {
                if v, ok := flow.DropReason_value[strings.ToUpper(reason)]; ok {
                        rs = append(rs, flowpb.DropReason(v))
                } else {
                        log.Warn("Ignoring invalid drop reason", logfields.Reason, reason)
                }
        }
```

We can then check in `ProcessFlow` with `slices.Contains(e.reasons, flow.GetDropReasonDesc())`.

devodev replied:
yes good call

41ks reviewed (approved):
Looks good to me. Thanks for taking this initiative!

kaworu reviewed (approved):
Thanks for the update @devodev
